waddrmgr/address.go

```diff
@@ -646,3 +646,154 @@ func newScriptAddress(m *ScopedKeyManager, account uint32, scriptHash,
 		scriptEncrypted: scriptEncrypted,
 	}, nil
 }
+
+// scriptAddress represents a pay-to-witness-script-hash address.
+type witnessScriptAddress struct {
+	manager         *ScopedKeyManager
+	account         uint32
+	address         *btcutil.AddressWitnessScriptHash
+	scriptEncrypted []byte
+	scriptCT        []byte
+	scriptMutex     sync.Mutex
+	used            bool
+}
+
+// Enforce scriptAddress satisfies the ManagedScriptAddress interface.
+var _ ManagedScriptAddress = (*witnessScriptAddress)(nil)
+
+// unlock decrypts and stores the associated script.  It will fail if the key is
+// invalid or the encrypted script is not available.  The returned clear text
+// script will always be a copy that may be safely used by the caller without
+// worrying about it being zeroed during an address lock.
+func (a *witnessScriptAddress) unlock(key EncryptorDecryptor) ([]byte, error) {
+	// Protect concurrent access to clear text script.
+	a.scriptMutex.Lock()
+	defer a.scriptMutex.Unlock()
+
+	if len(a.scriptCT) == 0 {
+		script, err := key.Decrypt(a.scriptEncrypted)
+		if err != nil {
+			str := fmt.Sprintf("failed to decrypt script for %s",
+				a.address)
+			return nil, managerError(ErrCrypto, str, err)
+		}
+
+		a.scriptCT = script
+	}
+
+	scriptCopy := make([]byte, len(a.scriptCT))
+	copy(scriptCopy, a.scriptCT)
+	return scriptCopy, nil
+}
+
+// lock zeroes the associated clear text private key.
+func (a *witnessScriptAddress) lock() {
+	// Zero and nil the clear text script associated with this address.
+	a.scriptMutex.Lock()
+	zero.Bytes(a.scriptCT)
+	a.scriptCT = nil
+	a.scriptMutex.Unlock()
+}
+
+// Account returns the account the address is associated with.  This will always
+// be the ImportedAddrAccount constant for script addresses.
+//
+// This is part of the ManagedAddress interface implementation.
+func (a *witnessScriptAddress) Account() uint32 {
```

Comment:
nit: would be nice if we could extract some of that duplicate code into a `baseScriptAddress` type that can be embedded into both types?